core/server/master/src/main/java/alluxio/master/meta/MetaDailyBackup.java

```diff
@@ -0,0 +1,181 @@
+import java.time.temporal.ChronoUnit;
+import java.util.TreeMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+
+/**
+ * Backing up primary master metadata everyday at a fixed UTC time.
+ */
+public final class MetaDailyBackup {
+  private static final Logger LOG = LoggerFactory.getLogger(MetaDailyBackup.class);
+  private static final long SHUTDOWN_TIMEOUT_MS = 5 * Constants.SECOND_MS;
+
+  private final String mBackupDir;
+  private final boolean mIsLocal;
+  private final int mRetainedFiles;
+  private final MetaMaster mMetaMaster;
+  private final ScheduledExecutorService mScheduledExecutor;
+  private final UnderFileSystem mUfs;
+
+  private ScheduledFuture<?> mBackup;
+
+  /**
+   * Constructs a new {@link MetaDailyBackup}.
+   *
+   * @param metaMaster the meta master
+   */
+  MetaDailyBackup(MetaMaster metaMaster, ScheduledExecutorService service, UnderFileSystem ufs) {
+    mMetaMaster = metaMaster;
+    mBackupDir = Configuration.get(PropertyKey.MASTER_BACKUP_DIRECTORY);
+    mRetainedFiles = Configuration.getInt(PropertyKey.MASTER_DAILY_BACKUP_FILES_RETAINED);
+    mScheduledExecutor = service;
+    mUfs = ufs;
+    mIsLocal = ufs.getUnderFSType().equals("local");
+  }
+
+  /**
+   * Starts {@link MetaDailyBackup}.
+   */
+  public void start() {
+    Preconditions.checkState(mBackup == null);
+    long delayedTimeInMillis = getTimeToNextBackup();
+    mBackup = mScheduledExecutor.scheduleAtFixedRate(this::dailyBackup,
+        delayedTimeInMillis, FormatUtils.parseTimeSize("1day"), TimeUnit.MILLISECONDS);
+    LOG.info("Daily metadata backup scheduled to start in {}",
+        CommonUtils.convertMsToClockTime(delayedTimeInMillis));
+  }
+
+  /**
+   * Gets the time gap between now and next backup time.
+   *
+   * @return the time gap to next backup
+   */
+  private long getTimeToNextBackup() {
+    LocalDateTime now = LocalDateTime.now(Clock.systemUTC());
+
+    DateTimeFormatter formatter = DateTimeFormatter.ofPattern("H:mm");
+    LocalTime backupTime = LocalTime.parse(Configuration
+        .get(PropertyKey.MASTER_DAILY_BACKUP_TIME), formatter);
+    LocalDateTime nextBackupTime = now.withHour(backupTime.getHour())
+        .withMinute(backupTime.getMinute());
+    if (nextBackupTime.isBefore(now)) {
+      nextBackupTime = nextBackupTime.plusDays(1);
+    }
+    return ChronoUnit.MILLIS.between(now, nextBackupTime);
+  }
+
+  /**
+   * The daily backup task.
+   */
+  private void dailyBackup() {
+    try {
+      BackupResponse resp = mMetaMaster.backup(new BackupOptions(mBackupDir, mIsLocal));
+      if (mIsLocal) {
+        LOG.info("Successfully backed up journal to {} on master {}",
+            resp.getBackupUri(), resp.getHostname());
+      } else {
+        LOG.info("Successfully backed up journal to {}", resp.getBackupUri());
+      }
+      try {
+        deleteOldestBackups();
```

Comment:
Moved outside. The inner try block is used because I think if we do not back up successfully, we could avoid deleting stale backup files.

If the logic is clearer, moving inner try block out also does not add many works.

Reply:
Thanks for introducing the better way! Changed to early return